README.md
## guh REST

## Installation

## Getting started

To start the web server simply type

    ./guh-webserver

You may configure the process with the following flags:

`ip`: The IP of the web server (default `0.0.0.0`)

`port`: The port of the web server (default `3000`)

`guh_ip`: The IP of the guhd server (default `127.0.0.1`)

`guh_port`: The port of the guhd server (default `1234`)

`static_folder`: The folder containing the static files of guh-webinterface (default: `./public`)

`conf_path`: A path pointing at a config file (default: `/etc/guh/guh-webserver.conf`)

For example:

    ./guh-webserver --guh_port=192.168.0.2 --guh_port=80

Here is an example for `guh-webserver.conf`:

    IP = "127.0.0.1"
    Port = 3000
    GuhIP = "192.168.0.3"
    GuhPort = 1234
    StaticFolder = "/my/cusom/public"

### License & Copyright

Copyright (c) 2015 guh

This software may be modified and distributed under the terms of the MIT license. See the LICENSE file for details.

### TODO

 - [ ] Refactor code and rename structs properly (Entity and EntityService)
 - [x] Properly handle deviceErrors etc. in responses
 - [ ] Refactor code to use something like "RegisterEndPoint()" to be able to autogenerate meta information (required params, relationship to guh core, etc.) about the API
 - [x] Convert ENV params to real command line params
   - [x] Params for guh-ip, guh-port, config-path, port
 - [x] Serve static files from configurable directory (http://stackoverflow.com/a/14187941/641032)


#### Endpoints

 - [x] get /core/introspect.json
   - [x] JSONRPC.Introspect
 - [x] get /core/version.json
   - [x] JSONRPC.Version
 - [x] get /devices.json
   - [x] Devices.GetConfiguredDevices
 - [x] get /devices/:id.json
   - [x] Devices.GetConfiguredDevices (filtered)
 - [x] delete /devices/:id.json
   - [x] Devices.RemoveConfiguredDevice
 - [x] post /devices.json
   - [x] Devices.AddConfiguredDevice
   - [x] Devices.PairDevice
 - [ ] post /devices/confirm_pairing.json
   - [ ] Devices.ConfirmPairing
 - [x] post /devices/:device_id/execute/:id.json
   - [x] Actions.ExecuteAction
 - [x] get /devices/:id/states.json
   - [x] Devices.GetStateValues
 - [ ] get /devices/:id/states/:id.json
   - [ ] Devices.GetStateValue
 - [x] get /device_classes.json
   - [x] Devices.GetSupportedDevices
 - [x] get /device_classes/:id.json
   - [x] Devices.GetSupportedDevices (filtered)
 - [x] get /device_classes/:device_class_id/action_types.json
   - [x] Devices.GetActionTypes
 - [ ] get /device_classes/:device_class_id/action_types/:id.json
   - [ ] Actions.GetActionType
 - [x] get /device_classes/:device_class_id/state_types.json
   - [x] Devices.GetStateTypes
 - [ ] get /device_classes/:device_class_id/state_types/:id.json
   - [ ] States.GetStateType
 - [x] get /device_classes/:device_class_id/event_types.json
   - [x] Devices.GetEventTypes
 - [ ] get /device_classes/:device_class_id/event_types/:id.json
   - [ ] Events.GetEventType
 - [x] get /device_classes/:id/discover.json
   - [x] Devices.GetDiscoveredDevices
 - [x] get /rules.json
   - [x] Rules.GetRules
 - [x] get /rules/:id.json
   - [x] Rules.GetRuleDetails
 - [x] post /rules.json
   - [x] Rules.AddRule
 - [x] patch /rules/:id/disable.json
   - [x] Rules.DisableRule
 - [x] patch /rules/:id/enable.json
   - [x] Rules.EnableRule
 - [x] delete /rules/:id.json
   - [x] Rules.RemoveRule
 - [x] get /vendors.json do
   - [x] Devices.GetSupportedVendors
 - [x] get /vendors/:id.json
   - [x] Devices.GetSupportedVendors (filtered)
 - [x] get /vendor/:id/device_classes.json
   - [x] Devices.GetSupportedDevices (filtered)
 - [ ] get /plugins.json
   - [ ] Devices.GetPlugins
 - [ ] get /plugins/:id.json
   - [ ] Devices.GetPluginConfiguration
 - [ ] put /devices/:id.json
   - [ ] Devices.SetPluginConfiguration
 - [ ] get /log.json
   - [ ] Logging.GetLogEntries
 - [ ] get /ws
   - [ ] JSONRPC.SetNotificationStatus


#### RPC

These RPCs are available but don't have a corresponding REST endpoint yet.

 - [ ] Rules.FindRules


debian/changelog
guh-webserver (0.0.1) unstable; urgency=medium

  * Initial release.

 -- Simon Stürz <simon.stuerz@guh.guru>  Mon, 26 Jan 2015 14:02:39 +0100


main.go
/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
 *                                                                                     *
 * Copyright (c) 2015 guh                                                              *
 *                                                                                     *
 * Permission is hereby granted, free of charge, to any person obtaining a copy        *
 * of this software and associated documentation files (the "Software"), to deal       *
 * in the Software without restriction, including without limitation the rights        *
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell           *
 * copies of the Software, and to permit persons to whom the Software is               *
 * furnished to do so, subject to the following conditions:                            *
 *                                                                                     *
 * The above copyright notice and this permission notice shall be included in all      *
 * copies or substantial portions of the Software.                                     *
 *                                                                                     *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR          *
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,            *
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE         *
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER              *
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,       *
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE       *
 * SOFTWARE.                                                                           *
 *                                                                                     *
 * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */

package main

import (
	"flag"
	"fmt"

	"github.com/go-martini/martini"
	"github.com/martini-contrib/render"
)

var IPFlag = flag.String("ip", "", "The ip address of the webserver (default: 0.0.0.0)")
var portFlag = flag.Int("port", 0, "The port of the webserver (default: 3000)")
var guhIPFlag = flag.String("guh_ip", "", "The IP address of the guh daemon (default: 127.0.0.1)")
var guhPortFlag = flag.Int("guh_port", 0, "The port of the guh daemon (default: 1234)")
var staticFolderFlag = flag.String("static_folder", "", "The location of the folder containing the static files (default: ./public)")
var confPathFlag = flag.String("conf_path", "/etc/guh/guh-webserver.conf", "The location of the config file (default: /etc/guh/guh-webserver.conf)")

func main() {
	// Parse the flag
	flag.Parse()

	config, guhConfig := runConfiguration()

	m := martini.Classic()
	m.Use(martini.Static(config.StaticFolder))
	m.Use(render.Renderer())

	DefineBaseEndPoints(m, guhConfig)
	DefineDeviceEndPoints(m, guhConfig)
	DefineDeviceClassEndPoints(m, guhConfig)
	DefineVendorEndPoints(m, guhConfig)
	DefineRuleEndPoints(m, guhConfig)

	m.RunOnAddr(fmt.Sprintf("%v:%v", config.IP, config.Port))
}
